proposals/0118-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,275 @@
+function is used to hash stake account addresses efficiently and uniformly
+across the blocks in the reward distribution phase. The hashing function for an
+epoch is created by seeding a new `SipHasher` with the parent block's blockhash.
+This hashing function can then be used to hash each active stake account's
+address into a u64 hash value. The reward distribution block index `I` can then
+be computed by applying the following formula to the hash:
+
+```
+I = (M * stake_address_hash) / 2^64
+```
+
+### Reward Distribution
+
+The reward distribution phase happens after the reward computation phase, ie.
+during the second block in the epoch as per the current design. It lasts for `M`
+blocks (see previous section). Rewards distribution for each block occurs before
+transaction processing occurs.
+
+#### `EpochRewards` Sysvar Account
+
+The `EpochRewards` sysvar account records whether the rewards distribution phase
+is in progress, as well as the details needed to resume distribution when
+starting from a snapshot during the reward distribution phase. These details
+include: the parameters needed to recalculate the reward partitions, the total
+rewards to be distributed, and the amount of rewards distributed so far.
+
+The layout of `EpochRewards` sysvar is shown in the following pseudo code.
+Prefixes indicate the offset of each field, and padding bytes are always zero.
+
+```
+#[repr(C)] // C representation, 8-byte alignment
+struct EpochRewards{
+   // whether the rewards period (calculation and distribution) is active
+   // byte. false: 0x00, true: 0x01
+   /* 0x00 */ active: bool,
+
+   // little-endian unsigned 64-bit integer
+   /* 0x08 */ distribution_starting_block_height: u64,
+   
+   // little-endian unsigned 64-bit integer
+   /* 0x10 */ num_partitions: u64,
+   
+   // byte-array of length 32
+   /* 0x18 */ parent_blockhash: Hash,
+
+   // total points calculated for the current epoch, where points equals the sum
+   // of delegated stake * credits observed for all delegations
+   // little-endian unsigned 128-bit integer
+   /* 0x38 */ total_points: u128, 
+
+   // total rewards for the current epoch, in lamports
+   // little-endian unsigned 64-bit integer
+   /* 0x48 */ total_rewards: u64,
+
+   // distributed rewards for the current epoch, in lamports
+   // little-endian unsigned 64-bit integer
+   /* 0x50 */ distributed_rewards: u64, 
+}
+```
+
+The `EpochRewards` sysvar is repopulated at the start of the first block of the
+epoch (before any transactions are processed), as both the total epoch rewards
+and vote account rewards become available at this time. The
+`distributed_rewards` field is updated per reward distribution for each block in
+the reward distribution phase (before transaction processing). The `active`
+field is set to false after the last partition is distributed (and
+`total_rewards == distributed_rewards`), ie. before processing transactions in
+block at height `X + M`.
+
+#### Booting from Snapshot
```

Comment:
Could we be a bit more verbose about what logic happens in this case? In an ideal world client teams should be able to use SIMDs, without looking at code, to implement new features. Thank-you! 🙏

Reply:
As discussed on Slack, lets update the SIMD with more precise implementation details after a POC has been implemented 😄